### DISCUSSION TITLE
A few questions about yash.

### DISCUSSION THREAD

**Discussion by vext01** [magicant/yash#36] [Q&A]
Hi,

As you've seen from the issues I've been raising, I've been experimenting with yash. I'm coming from fish, which I liked, but ultimately got annoyed with the syntax deviations from what I'm used to and the `set` system it uses.

I had a few questions:
 - Is there a way to re-run the last command? I would use `<up>, <enter>`, but due to shared history, that isn't necessarily the last command. In bash and zsh, `!!` would work, but this doesn't work on yash (or sh, ksh and fish).
 
 - Is there any plan to add something like `beginning-search-backward`, but which matches the input *anywhere* in the history lines (without the string being anchored to the beginning of the line). For example, in fish if I type `ls` and press up repeatedly it could offer history lines like `ls /tmp`, `rm rails`, `man ls` on account of them all *containing* the substring `ls`.

 - Is there a way to copy the currently matched history line (when using <ctrl+r> and <ctrl+s>) into an editable line for amending?

Many thanks for yash. I like it!

**Comment by magicant** [comment]
> Is there a way to re-run the last command? I would use `<up>, <enter>`, but due to shared history, that isn't necessarily the last command. In bash and zsh, `!!` would work, but this doesn't work on yash (or sh, ksh and fish).

Separate history is not supported and has long been a pending issue (#25). 

Setting a different `$HISTFILE` for each session makes the history separate, but you need to merge the file manually.

> Is there any plan to add something like `beginning-search-backward`, but which matches the input _anywhere_ in the history lines (without the string being anchored to the beginning of the line). For example, in fish if I type `ls` and press up repeatedly it could offer history lines like `ls /tmp`, `rm rails`, `man ls` on account of them all _containing_ the substring `ls`.

Implementing that behavior will not be as easy as `beginning-search-backward`.
For now, repeating `<ctrl+r>` produces a similar effect.

> Is there a way to copy the currently matched history line (when using `<ctrl+r>` and `<ctrl+s>`) into an editable line for amending?

`<esc>` or `<ctrl+j>`

**Comment by vext01** [comment]
Hi,

Thanks for the reply!

> Separate history is not supported and has long been a pending issue (https://github.com/magicant/yash/issues/25).
>
> Setting a different $HISTFILE for each session makes the history separate, but you need to merge the file manually.

Yeah, i realised I could use such a hack, but it's not obvious to me how to merge the histories in a sane and automated fashion. Maybe an EXIT trap?

> Implementing that behavior will not be as easy as beginning-search-backward.
> For now, repeating <ctrl+r> produces a similar effect.

Yes. The only problem is, CTRL+r doesn't inherit the text from the line edit buffer, so you have to type your search term again. Maybe we could add a variant of the search command which copies the line edit buffer into the search buffer?

I could even have a go at a pull request for that if you think it would be acceptable.

> <esc> or <ctrl+j>

Perfect! Thank you!

Is this documented? I don't see it when grepping the manual `escape`, `Escape` and `\e`.

**Reply by magicant** [reply]
> it's not obvious to me how to merge the histories in a sane and automated fashion. Maybe an EXIT trap?

Unfortunately, it is known that there is no simple method to merge history files correctly and automatically... Someone tried to find one in the past in vain, and I don't have a good idea either.

> The only problem is, CTRL+r doesn't inherit the text from the line edit buffer, so you have to type your search term again. Maybe we could add a variant of the search command which copies the line edit buffer into the search buffer? I could even have a go at a pull request for that if you think it would be acceptable.

Sounds nice!

> Is this documented? I don't see it when grepping the manual escape, Escape and \e.

It appears as the `srch-accept-search` command in <https://magicant.github.io/yash/doc/lineedit.html#search-commands>.

**Comment by vext01** [comment]
> Separate history is not supported and has long been a pending issue

One other comment. Often I don't need full history and just want to run the last command again. This could be achieved with the `!!` builtin of (e.g.) bash or zsh.

Would you object to adding `!!` to yash?

**Reply by magicant** [reply]
You might find this alias useful, which is included in the [default yashrc](https://github.com/magicant/yash/blob/bd1fd3f1420569ce09551cfc6a8acfb9092fd6dc/share/initialization/common#L94).

`alias r='fc -s'`

or if you prefer to name it `!!` then

`alias !!='fc -s'`

**Reply by alterae** [reply]
Apologies for jumping in on this very late but I managed to fairly trivially reproduce most of the `!!` functionality I'm used to from zsh using a global alias.

```sh
alias -g !!='$(fc -ln | tail -2 | head -1)'
```

This allows both simple rerunning of the previous command, as well as prepending to the previous command and rerunning, at the cost of not being POSIX-compatible. It also does not support echoing the command before rerunning—I implemented that for `sudo` specifically using a wrapper function.

```sh
function sudo {
	echo sudo $@
	command -e sudo $@
}
```

Overall, it works more or less as expected.

```sh
$ echo hello
hello
$ !!
hello
$ whoami
chell
$ sudo !!
sudo whoami
root
```
